apollo-runtime/src/main/java/com/apollographql/android/HttpOperationRequest.java

```diff
@@ -0,0 +1,121 @@
+  private static final MediaType MEDIA_TYPE = MediaType.parse("application/graphql; charset=utf-8");
+
+  private final Operation operation;
+  private final Moshi moshi;
+  private final Request baseRequest;
+  private final okhttp3.Call.Factory callFactory;
+  private final ResponseBodyConverter responseBodyConverter;
+  private final AtomicReference<Call> callRef = new AtomicReference<>(null);
+
+  HttpOperationRequest(Operation operation, Moshi moshi, okhttp3.Call.Factory callFactory, Request baseRequest,
+      ResponseFieldMapper responseFieldMapper, Map<ScalarType, CustomTypeAdapter> customTypeAdapters) {
+    this.operation = operation;
+    this.moshi = moshi;
+    this.callFactory = callFactory;
+    this.baseRequest = baseRequest;
+    responseBodyConverter = new ResponseBodyConverter(operation, responseFieldMapper, customTypeAdapters);
+  }
+
+  @Override public void cancel() {
+    Call httpCall = callRef.get();
+    if (httpCall != null) {
+      httpCall.cancel();
+    }
+  }
+
+  @Override public Response<T> execute() throws IOException {
+    RequestBody requestBody = requestBody(operation);
+    Request request = baseRequest.newBuilder()
+        .post(requestBody)
+        .build();
+
+    Call call = callFactory.newCall(request);
+    if (callRef.compareAndSet(null, call)) {
+      okhttp3.Response response = call.execute();
+      return parse(response);
+    } else {
+      throw new IllegalStateException("Already Executed");
+    }
+  }
+
+  @Override public OperationRequest<T> enqueue(final Callback<T> callback) {
+    Request request;
+    try {
+      RequestBody requestBody = requestBody(operation);
+      request = baseRequest.newBuilder()
+          .post(requestBody)
+          .build();
+    } catch (Exception e) {
+      if (callback != null) {
+        callback.onFailure(e);
+      }
+      return this;
+    }
+
+    Call httpCall = callFactory.newCall(request);
+    if (callRef.compareAndSet(null, httpCall)) {
+      httpCall.enqueue(new okhttp3.Callback() {
+        @Override public void onFailure(Call call, IOException e) {
+          if (callback != null) {
+            callback.onFailure(e);
+          }
+        }
+
+        @Override public void onResponse(Call call, okhttp3.Response httpResponse) throws IOException {
+          Response<T> response;
+          try {
+            response = parse(httpResponse);
+          } catch (Exception e) {
+            if (callback != null) {
+              callback.onFailure(e);
+            }
+            return;
+          }
+
+          if (callback != null) {
+            callback.onResponse(response);
+          }
+        }
+      });
+      return this;
+    } else {
+      throw new IllegalStateException("Already Executed");
+    }
+  }
+
+  private Response<T> parse(okhttp3.Response response) throws IOException {
+    int code = response.code();
+    if (code < 200 || code >= 300) {
```

Comment:
you probably could use:  https://square.github.io/okhttp/3.x/okhttp/okhttp3/Response.html#isSuccessful--